src/type.cpp

```diff
@@ -152,3 +152,64 @@ std::unique_ptr<Type> FuncType::Clone() const {
   return std::make_unique<FuncType>(std::move(cloned_return_type),
                                     std::move(cloned_param_types));
 }
+
+bool StructType::IsEqual(const Type& that) const noexcept {
+  if (const auto* that_struct = dynamic_cast<const StructType*>(&that)) {
+    if (that_struct->size() != that.size()) {
+      return false;
+    }
+    for (auto i = std::size_t{0}, e = field_types_.size(); i < e; ++i) {
+      if (!that_struct->field_types_.at(i)->IsEqual(*field_types_.at(i))) {
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+std::size_t StructType::size() const {
+  // TODO: There may be unnamed padding at the end of a structure or union.
+  auto size = std::size_t{0};
+  for (auto i = std::size_t{0}, e = field_types_.size(); i < e; ++i) {
+    size += field_types_.at(i)->size();
+  }
```

Comment:
Considering we're iterating over the entire vector, I'd suggest using a range-based loop.